File: 0874-backspace-string-compare/0874-backspace-string-compare.go
func backspaceCompare(s string, t string) bool { 
    i, j := len(s) - 1, len(t) - 1

    for {
        bs := 0

        for i >= 0 && (s[i] == '#' || bs > 0) {
            if s[i] == '#' {
                bs++
                i--
            } else {
                i--
                bs--
            }
        }

        bs = 0

        for j >= 0 && (t[j] == '#' || bs > 0) {
            if t[j] == '#' {
                bs++
                j--
            } else {
                j--
                bs--
            }
        }

        if i < 0 || j < 0 { 
            break
        }

        if s[i] != t[j] {
            break
        }

        i--
        j--
    }

    return i < 0 && j < 0
}

File: 0874-backspace-string-compare/0874-backspace-string-compare.rb
# @param {String} s
# @param {String} t
# @return {Boolean}
def backspace_compare(s, t)
    return backspace(s) == backspace(t)
end

def backspace(s)
    bs = 0
    s2 = ""
    s.chars.reverse_each do |c|
        if c == '#'
            bs += 1
            next
        end
        
        if bs > 0
            bs -= 1
            next
        end

        s2 << c
    end

    s2.reverse
end